### PR TITLE
M192 : T383 : A001: implement data product event retrieve 1

### PR DESCRIPTION
**Pull request by rajvikram**
@MauriceManning

Added a note for Maurice on where to add the code in event_poll.py

### CODE REVIEW

File: ion/services/sa/product/data_product_management_service.py [hunk 1]
```diff
@@ -288,7 +291,63 @@ def get_data_product_updates(self, data_product_id_list=None, since_timestamp=""
...
+            filter_mindate = dqb.gte(DQ.RA_TS_CREATED, since_timestamp)
+            where = dqb.and_(filter_origins, filter_types, filter_mindate)
+
+            order_by = dqb.order_by([["ts_created", "desc"]])  # Descending order by time
+            dqb.build_query(where=where, order_by=order_by, limit=10, skip=0, id_only=False)
+            query = dqb.get_query()
+
+            event_objs = self.container.event_repository.event_store.find_by_query(query)
+
+        except:
```
Comment (mmeisinger) [inline ion/services/sa/product/data_product_management_service.py:317]: Please no unqualified excepts. If necessary use except Exception

File: ion/services/sa/product/data_product_management_service.py [hunk 2]
```diff
@@ -288,7 +291,63 @@ def get_data_product_updates(self, data_product_id_list=None, since_timestamp=""
...
+        try:
+            dqb = DatastoreQueryBuilder(datastore=DataStore.DS_EVENTS, profile=DataStore.DS_PROFILE.EVENTS)
+
+            filter_origins = dqb.in_(DQ.EA_ORIGIN, *data_product_id_list)
+            filter_types = dqb.in_(DQ.ATT_TYPE, "ResourceModifiedEvent")
+            filter_mindate = dqb.gte(DQ.RA_TS_CREATED, since_timestamp)
+            where = dqb.and_(filter_origins, filter_types, filter_mindate)
+
+            order_by = dqb.order_by([["ts_created", "desc"]])  # Descending order by time
+            dqb.build_query(where=where, order_by=order_by, limit=10, skip=0, id_only=False)
```
Comment (mmeisinger) [inline ion/services/sa/product/data_product_management_service.py:312]: A limit of 10 may not get you where you want

File: ion/services/sa/product/data_product_management_service.py [hunk 3]
```diff
@@ -288,7 +291,63 @@ def get_data_product_updates(self, data_product_id_list=None, since_timestamp=""
...
+
+        # Start populating the response structure
+        for data_product_id in data_product_id_list:
+            response_data[data_product_id] = {"dataset_id" : None,
+                                              "updated" : False,
+                                              "current_geospatial_bounds" : None,
+                                              "current_temporal_bounds" : None}
+
+            # Need dataset id in response data
+            ds_ids,_ = self.clients.resource_registry.find_objects(subject=data_product_id,
```
Comment (mmeisinger) [inline ion/services/sa/product/data_product_management_service.py:329]: This is a RR call inside a for loop. There should be a warning note that this should be replaced with more efficient code.

File: ion/util/dataproducts_to_monitor.txt [hunk 4]
```diff
@@ -0,0 +1,2 @@
+de57a11b812b4964af3157b76447ec8a
```
Comment (mmeisinger) [inline ion/util/dataproducts_to_monitor.txt:1]: Not sure what this file is - it should not be part of this PR

File: ion/util/event_poll.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
+import sys
+import os.path
+import ast
+import json, time
+import urllib, urllib2
+
+dataproducts_to_monitor_file = "dataproducts_to_monitor.txt"
```
Comment (mmeisinger) [inline ion/util/event_poll.py:7]: You cannot put code like this into ion.util, which are general purposed utilities. You don't need to poll either. Just wait the 1.1s max until the event persister has persisted the events.

File: ion/util/dataproducts_to_monitor.txt (hunk 4 above)
Comment (MauriceManning) [inline ion/util/dataproducts_to_monitor.txt:1]: Put this as an example of how to set up the file in the documentation section of the script below which I would call something like data_product_monitor_and_retrieve.py.

File: ion/services/sa/product/data_product_management_service.py [hunk 6]
```diff
@@ -288,7 +291,63 @@ def get_data_product_updates(self, data_product_id_list=None, since_timestamp=""
...
+        if data_product_id_list == None or len(data_product_id_list) == 0:
+            raise  BadRequest("Please pass a valid data_product_id")
+
+        # Build query structure for fetching events for the data products passed
+        try:
+            dqb = DatastoreQueryBuilder(datastore=DataStore.DS_EVENTS, profile=DataStore.DS_PROFILE.EVENTS)
+
+            filter_origins = dqb.in_(DQ.EA_ORIGIN, *data_product_id_list)
+            filter_types = dqb.in_(DQ.ATT_TYPE, "ResourceModifiedEvent")
+            filter_mindate = dqb.gte(DQ.RA_TS_CREATED, since_timestamp)
```
Comment (MauriceManning) [inline ion/services/sa/product/data_product_management_service.py:308]: what time format is expected by the query here? a unix timestamp?
I would like the since_timestamp to be ISO format so that it is readable, then convert as required.

File: ion/services/sa/product/data_product_management_service.py (hunk 2 above)
Comment (rajvikram) [inline ion/services/sa/product/data_product_management_service.py:312]: Interesting .. I could have sworn I set this to 100,000

File: ion/services/sa/product/data_product_management_service.py (hunk 3 above)
Comment (rajvikram) [inline ion/services/sa/product/data_product_management_service.py:329]: I did try using the find_objects_mult and it returned me a list of ids .. I
wasn't sure which object id corresponds to which subject or was I supposed
to assume it was the same order as the input subject id list. Since we just
wanted the code working for now so that Maurice could made progress, I went
for what I knew will work. Optimizations to follow.

File: ion/util/dataproducts_to_monitor.txt (hunk 4 above)
Comment (rajvikram) [inline ion/util/dataproducts_to_monitor.txt:1]: This is what I was told to do. The file contains a list of ids to be polled
for event updates.

File: ion/util/event_poll.py (hunk 5 above)
Comment (rajvikram) [inline ion/util/event_poll.py:7]: Again these were the instructions I had. To put the script in ion/util and
have it read a dataproducts_to_monitor.txt file

File: ion/util/dataproducts_to_monitor.txt (hunk 4 above)
Comment (rajvikram) [inline ion/util/dataproducts_to_monitor.txt:1]: Ok I'll change the name of the polling script.

File: ion/util/event_poll.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    with open(dataproducts_to_monitor_file) as f:
+        data_product_ids = f.read().splitlines()
+
+
+    # Now that we have the list of data product IDs, get events for each
+    # Poll every few seconds
+    url = server_addr + "/ion-service/data_product_management/get_data_product_updates"
+    while True:
+
+        parameters = {"data_product_id_list" : data_product_ids,
```
Comment (MauriceManning) [inline ion/util/event_poll.py:39]: can you use a ISO 8601 timestamp here for ease of use?

Comment (rajvikram) [inline ion/util/event_poll.py:39]: Will do.

File: ion/services/sa/product/data_product_management_service.py (hunk 1 above)
Comment (rajvikram) [inline ion/services/sa/product/data_product_management_service.py:317]: Got it .. I'll make the changes.

File: ion/util/event_poll.py (hunk 5 above)
Comment (rajvikram) [inline ion/util/event_poll.py:7]: Michael, are you saying this script should not be in ion.util ? Or that the hard coded value of dataproducts_to_monitor_file is the issue ? I've removed the hard code and now the users have to specify the file from the command line.

Also Maurice and I are not sure by what you mean by not needing to poll. Please grab me on chat or Skype. Thanks.
